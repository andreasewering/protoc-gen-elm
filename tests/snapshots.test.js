const cp = require("child_process");
const path = require("path");
const fs = require("fs");
const { promisify } = require("util");

const protoPath = path.join(__dirname, "proto");
const generatedPath = path.join(__dirname, "..", "generated");

const cases = [
  { input: "single_enum.proto", output: "Proto/SingleEnum.elm" },
  { input: "basic_message.proto", output: "Proto/BasicMessage.elm" },
  { input: "oneof.proto", output: "Proto/Oneof.elm" },
  {
    input: ["imported.proto", "importing.proto"],
    output: ["Proto/Importing.elm", "Proto/Imported.elm"],
  },
  { input: "multiple_oneof.proto", output: "Proto/MultipleOneof.elm" },
  { input: "package.proto", output: "Any/Package.elm" },
];

const exec = (command) =>
  new Promise((resolve, reject) => {
    cp.exec(command, (err) => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });

const readFile = (filePath) =>
  promisify(fs.readFile)(filePath, { encoding: "utf-8" });

describe("protoc-plugin-elm", () => {
  beforeAll((done) => {
    fs.rm(generatedPath, { recursive: true, force: true }, () => {
      fs.mkdir(generatedPath, { recursive: true }, done);
    });
  });

  test.each(cases)(
    "generates expected code for $input",
    async ({ input, output }) => {
      console.log(`Starting ${input} generation`);
      const args = Array.isArray(input) ? input.join(" ") : input;
      await exec(
        `protoc --plugin="protoc-gen-elm=${path.resolve(
          __dirname,
          "..",
          "index.js"
        )}" --proto_path=${protoPath} --elm_out=${generatedPath} ${args}`
      );

      const outputFilenames = Array.isArray(output) ? output : [output];
      // check that the generated files compile
      await exec(
        `elm make ${outputFilenames
          .map((filename) => path.join(generatedPath, filename))
          .join(" ")}`
      );

      // compare with previously generated file as a regression test
      const generatedContent = await Promise.all(
        outputFilenames.map(async (filename) => {
          const outputPath = path.join(generatedPath, filename);
          return readFile(outputPath);
        })
      );
      expect(generatedContent.join("\n")).toMatchSnapshot();
    }
  );
});
