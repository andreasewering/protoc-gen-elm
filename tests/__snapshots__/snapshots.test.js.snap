// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protoc-plugin-elm generates expected code for ["imported.proto", "importing.proto"] 1`] = `
"{- !!! DO NOT EDIT THIS FILE MANUALLY !!! -}


module Proto.Importing exposing
    ( Nested
    , nestedDecoder
    , toNestedEncoder
    )

{-| ProtoBuf module: \`Proto.Importing\`

This module was generated automatically using

  - [\`protoc-gen-elm\`](https://www.npmjs.com/package/protoc-gen-elm) 1.0.0-beta-2
  - \`protoc\` 3.14.0
  - the following specification file: \`importing.proto\`

To run it use [\`elm-protocol-buffers\`](https://package.elm-lang.org/packages/eriktim/elm-protocol-buffers/1.1.0) version 1.1.0 or higher.


# Model

@docs Nested


# Decoder

@docs nestedDecoder


# Encoder

@docs toNestedEncoder

-}

import Proto.Imported exposing (..)
import Protobuf.Decode as Decode
import Protobuf.Encode as Encode



-- MODEL


{-| \`Nested\` message
-}
type alias Nested =
    { normalProperty : String
    , nestedProperty : Maybe Imported
    }



-- DECODER


{-| \`Nested\` decoder
-}
nestedDecoder : Decode.Decoder Nested
nestedDecoder =
    Decode.message (Nested \\"\\" Nothing)
        [ Decode.optional 1 Decode.string setNormalProperty
        , Decode.optional 2 (Decode.map Just importedDecoder) setNestedProperty
        ]



-- ENCODER


{-| \`Nested\` encoder
-}
toNestedEncoder : Nested -> Encode.Encoder
toNestedEncoder model =
    Encode.message
        [ ( 1, Encode.string model.normalProperty )
        , ( 2, (Maybe.withDefault Encode.none << Maybe.map toImportedEncoder) model.nestedProperty )
        ]



-- SETTERS


setNormalProperty : a -> { b | normalProperty : a } -> { b | normalProperty : a }
setNormalProperty value model =
    { model | normalProperty = value }


setNestedProperty : a -> { b | nestedProperty : a } -> { b | nestedProperty : a }
setNestedProperty value model =
    { model | nestedProperty = value }

{- !!! DO NOT EDIT THIS FILE MANUALLY !!! -}


module Proto.Imported exposing
    ( Imported
    , importedDecoder
    , toImportedEncoder
    )

{-| ProtoBuf module: \`Proto.Imported\`

This module was generated automatically using

  - [\`protoc-gen-elm\`](https://www.npmjs.com/package/protoc-gen-elm) 1.0.0-beta-2
  - \`protoc\` 3.14.0
  - the following specification file: \`imported.proto\`

To run it use [\`elm-protocol-buffers\`](https://package.elm-lang.org/packages/eriktim/elm-protocol-buffers/1.1.0) version 1.1.0 or higher.


# Model

@docs Imported


# Decoder

@docs importedDecoder


# Encoder

@docs toImportedEncoder

-}

import Protobuf.Decode as Decode
import Protobuf.Encode as Encode



-- MODEL


{-| \`Imported\` message
-}
type alias Imported =
    { first : String
    , second : Bool
    }



-- DECODER


{-| \`Imported\` decoder
-}
importedDecoder : Decode.Decoder Imported
importedDecoder =
    Decode.message (Imported \\"\\" False)
        [ Decode.optional 1 Decode.string setFirst
        , Decode.optional 2 Decode.bool setSecond
        ]



-- ENCODER


{-| \`Imported\` encoder
-}
toImportedEncoder : Imported -> Encode.Encoder
toImportedEncoder model =
    Encode.message
        [ ( 1, Encode.string model.first )
        , ( 2, Encode.bool model.second )
        ]



-- SETTERS


setFirst : a -> { b | first : a } -> { b | first : a }
setFirst value model =
    { model | first = value }


setSecond : a -> { b | second : a } -> { b | second : a }
setSecond value model =
    { model | second = value }
"
`;

exports[`protoc-plugin-elm generates expected code for basic_message.proto 1`] = `
"{- !!! DO NOT EDIT THIS FILE MANUALLY !!! -}


module Proto.BasicMessage exposing
    ( BasicMessage
    , basicMessageDecoder
    , toBasicMessageEncoder
    )

{-| ProtoBuf module: \`Proto.BasicMessage\`

This module was generated automatically using

  - [\`protoc-gen-elm\`](https://www.npmjs.com/package/protoc-gen-elm) 1.0.0-beta-2
  - \`protoc\` 3.14.0
  - the following specification file: \`basic_message.proto\`

To run it use [\`elm-protocol-buffers\`](https://package.elm-lang.org/packages/eriktim/elm-protocol-buffers/1.1.0) version 1.1.0 or higher.


# Model

@docs BasicMessage


# Decoder

@docs basicMessageDecoder


# Encoder

@docs toBasicMessageEncoder

-}

import Protobuf.Decode as Decode
import Protobuf.Encode as Encode



-- MODEL


{-| \`BasicMessage\` message
-}
type alias BasicMessage =
    { stringProperty : String
    , intProperty : Int
    , floatProperty : Float
    , boolProperty : Bool
    }



-- DECODER


{-| \`BasicMessage\` decoder
-}
basicMessageDecoder : Decode.Decoder BasicMessage
basicMessageDecoder =
    Decode.message (BasicMessage \\"\\" 0 0 False)
        [ Decode.optional 1 Decode.string setStringProperty
        , Decode.optional 2 Decode.int32 setIntProperty
        , Decode.optional 3 Decode.float setFloatProperty
        , Decode.optional 4 Decode.bool setBoolProperty
        ]



-- ENCODER


{-| \`BasicMessage\` encoder
-}
toBasicMessageEncoder : BasicMessage -> Encode.Encoder
toBasicMessageEncoder model =
    Encode.message
        [ ( 1, Encode.string model.stringProperty )
        , ( 2, Encode.int32 model.intProperty )
        , ( 3, Encode.float model.floatProperty )
        , ( 4, Encode.bool model.boolProperty )
        ]



-- SETTERS


setStringProperty : a -> { b | stringProperty : a } -> { b | stringProperty : a }
setStringProperty value model =
    { model | stringProperty = value }


setIntProperty : a -> { b | intProperty : a } -> { b | intProperty : a }
setIntProperty value model =
    { model | intProperty = value }


setFloatProperty : a -> { b | floatProperty : a } -> { b | floatProperty : a }
setFloatProperty value model =
    { model | floatProperty = value }


setBoolProperty : a -> { b | boolProperty : a } -> { b | boolProperty : a }
setBoolProperty value model =
    { model | boolProperty = value }
"
`;

exports[`protoc-plugin-elm generates expected code for multiple_oneof.proto 1`] = `
"{- !!! DO NOT EDIT THIS FILE MANUALLY !!! -}


module Proto.MultipleOneof exposing
    ( Oneof1Msg(..), Oneof1, Oneof2Msg(..), Oneof2
    , oneof1Decoder, oneof2Decoder
    , toOneof1Encoder, toOneof2Encoder
    )

{-| ProtoBuf module: \`Proto.MultipleOneof\`

This module was generated automatically using

  - [\`protoc-gen-elm\`](https://www.npmjs.com/package/protoc-gen-elm) 1.0.0-beta-2
  - \`protoc\` 3.14.0
  - the following specification file: \`multiple_oneof.proto\`

To run it use [\`elm-protocol-buffers\`](https://package.elm-lang.org/packages/eriktim/elm-protocol-buffers/1.1.0) version 1.1.0 or higher.


# Model

@docs Oneof1Msg, Oneof1, Oneof2Msg, Oneof2


# Decoder

@docs oneof1Decoder, oneof2Decoder


# Encoder

@docs toOneof1Encoder, toOneof2Encoder

-}

import Protobuf.Decode as Decode
import Protobuf.Encode as Encode



-- MODEL


{-| Oneof1Msg
-}
type Oneof1Msg
    = MsgOptionA String
    | MsgOptionB String


{-| \`Oneof1\` message
-}
type alias Oneof1 =
    { msg : Maybe Oneof1Msg
    }


{-| Oneof2Msg
-}
type Oneof2Msg
    = MsgOptionC String
    | MsgOptionD String


{-| \`Oneof2\` message
-}
type alias Oneof2 =
    { msg : Maybe Oneof2Msg
    }



-- DECODER


{-| \`Oneof1\` decoder
-}
oneof1Decoder : Decode.Decoder Oneof1
oneof1Decoder =
    Decode.message (Oneof1 Nothing)
        [ Decode.oneOf
            [ ( 1, Decode.map MsgOptionA Decode.string )
            , ( 2, Decode.map MsgOptionB Decode.string )
            ]
            setMsg
        ]


{-| \`Oneof2\` decoder
-}
oneof2Decoder : Decode.Decoder Oneof2
oneof2Decoder =
    Decode.message (Oneof2 Nothing)
        [ Decode.oneOf
            [ ( 1, Decode.map MsgOptionC Decode.string )
            , ( 2, Decode.map MsgOptionD Decode.string )
            ]
            setMsg
        ]



-- ENCODER


toOneof1MsgEncoder : Oneof1Msg -> ( Int, Encode.Encoder )
toOneof1MsgEncoder model =
    case model of
        MsgOptionA value ->
            ( 1, Encode.string value )

        MsgOptionB value ->
            ( 2, Encode.string value )


{-| \`Oneof1\` encoder
-}
toOneof1Encoder : Oneof1 -> Encode.Encoder
toOneof1Encoder model =
    Encode.message
        [ Maybe.withDefault ( 0, Encode.none ) <| Maybe.map toOneof1MsgEncoder model.msg
        ]


toOneof2MsgEncoder : Oneof2Msg -> ( Int, Encode.Encoder )
toOneof2MsgEncoder model =
    case model of
        MsgOptionC value ->
            ( 1, Encode.string value )

        MsgOptionD value ->
            ( 2, Encode.string value )


{-| \`Oneof2\` encoder
-}
toOneof2Encoder : Oneof2 -> Encode.Encoder
toOneof2Encoder model =
    Encode.message
        [ Maybe.withDefault ( 0, Encode.none ) <| Maybe.map toOneof2MsgEncoder model.msg
        ]



-- SETTERS


setMsg : a -> { b | msg : a } -> { b | msg : a }
setMsg value model =
    { model | msg = value }
"
`;

exports[`protoc-plugin-elm generates expected code for oneof.proto 1`] = `
"{- !!! DO NOT EDIT THIS FILE MANUALLY !!! -}


module Proto.Oneof exposing
    ( OneOfMsg(..), OneOf
    , oneOfDecoder
    , toOneOfEncoder
    )

{-| ProtoBuf module: \`Proto.Oneof\`

This module was generated automatically using

  - [\`protoc-gen-elm\`](https://www.npmjs.com/package/protoc-gen-elm) 1.0.0-beta-2
  - \`protoc\` 3.14.0
  - the following specification file: \`oneof.proto\`

To run it use [\`elm-protocol-buffers\`](https://package.elm-lang.org/packages/eriktim/elm-protocol-buffers/1.1.0) version 1.1.0 or higher.


# Model

@docs OneOfMsg, OneOf


# Decoder

@docs oneOfDecoder


# Encoder

@docs toOneOfEncoder

-}

import Protobuf.Decode as Decode
import Protobuf.Encode as Encode



-- MODEL


{-| OneOfMsg
-}
type OneOfMsg
    = MsgAString String
    | MsgAnInt Int
    | MsgABool Bool


{-| \`OneOf\` message
-}
type alias OneOf =
    { msg : Maybe OneOfMsg
    }



-- DECODER


{-| \`OneOf\` decoder
-}
oneOfDecoder : Decode.Decoder OneOf
oneOfDecoder =
    Decode.message (OneOf Nothing)
        [ Decode.oneOf
            [ ( 1, Decode.map MsgAString Decode.string )
            , ( 2, Decode.map MsgAnInt Decode.int32 )
            , ( 3, Decode.map MsgABool Decode.bool )
            ]
            setMsg
        ]



-- ENCODER


toOneOfMsgEncoder : OneOfMsg -> ( Int, Encode.Encoder )
toOneOfMsgEncoder model =
    case model of
        MsgAString value ->
            ( 1, Encode.string value )

        MsgAnInt value ->
            ( 2, Encode.int32 value )

        MsgABool value ->
            ( 3, Encode.bool value )


{-| \`OneOf\` encoder
-}
toOneOfEncoder : OneOf -> Encode.Encoder
toOneOfEncoder model =
    Encode.message
        [ Maybe.withDefault ( 0, Encode.none ) <| Maybe.map toOneOfMsgEncoder model.msg
        ]



-- SETTERS


setMsg : a -> { b | msg : a } -> { b | msg : a }
setMsg value model =
    { model | msg = value }
"
`;

exports[`protoc-plugin-elm generates expected code for package.proto 1`] = `
"{- !!! DO NOT EDIT THIS FILE MANUALLY !!! -}


module Any.Package exposing
    ( Test
    , testDecoder
    , toTestEncoder
    )

{-| ProtoBuf module: \`Any.Package\`

This module was generated automatically using

  - [\`protoc-gen-elm\`](https://www.npmjs.com/package/protoc-gen-elm) 1.0.0-beta-2
  - \`protoc\` 3.14.0
  - the following specification file: \`package.proto\`

To run it use [\`elm-protocol-buffers\`](https://package.elm-lang.org/packages/eriktim/elm-protocol-buffers/1.1.0) version 1.1.0 or higher.


# Model

@docs Test


# Decoder

@docs testDecoder


# Encoder

@docs toTestEncoder

-}

import Protobuf.Decode as Decode
import Protobuf.Encode as Encode



-- MODEL


{-| \`Test\` message
-}
type alias Test =
    {}



-- DECODER


{-| \`Test\` decoder
-}
testDecoder : Decode.Decoder Test
testDecoder =
    Decode.message Test
        []



-- ENCODER


{-| \`Test\` encoder
-}
toTestEncoder : Test -> Encode.Encoder
toTestEncoder _ =
    Encode.message
        []



-- SETTERS



"
`;

exports[`protoc-plugin-elm generates expected code for single_enum.proto 1`] = `
"{- !!! DO NOT EDIT THIS FILE MANUALLY !!! -}


module Proto.SingleEnum exposing
    ( AnEnum(..)
    , anEnumDecoder
    , toAnEnumEncoder
    )

{-| ProtoBuf module: \`Proto.SingleEnum\`

This module was generated automatically using

  - [\`protoc-gen-elm\`](https://www.npmjs.com/package/protoc-gen-elm) 1.0.0-beta-2
  - \`protoc\` 3.14.0
  - the following specification file: \`single_enum.proto\`

To run it use [\`elm-protocol-buffers\`](https://package.elm-lang.org/packages/eriktim/elm-protocol-buffers/1.1.0) version 1.1.0 or higher.


# Model

@docs AnEnum


# Decoder

@docs anEnumDecoder


# Encoder

@docs toAnEnumEncoder

-}

import Protobuf.Decode as Decode
import Protobuf.Encode as Encode



-- MODEL


{-| \`AnEnum\` enumeration
-}
type AnEnum
    = OptionA
    | OptionB
    | OPTIONC
    | AnEnumUnrecognized_ Int



-- DECODER


anEnumDecoder : Decode.Decoder AnEnum
anEnumDecoder =
    Decode.int32
        |> Decode.map
            (\\\\value ->
                case value of
                    0 ->
                        OptionA

                    1 ->
                        OptionB

                    2 ->
                        OPTIONC

                    v ->
                        AnEnumUnrecognized_ v
            )



-- ENCODER


toAnEnumEncoder : AnEnum -> Encode.Encoder
toAnEnumEncoder value =
    Encode.int32 <|
        case value of
            OptionA ->
                0

            OptionB ->
                1

            OPTIONC ->
                2

            AnEnumUnrecognized_ v ->
                v



-- SETTERS



"
`;
